src/Tokenizer.cc

```diff
@@ -468,15 +469,47 @@ namespace onmt
         return;
 
       std::vector<std::string> chunks = unicode::split_utf8(text, " ");
-      for (const auto& chunk: chunks)
+      for (auto& chunk: chunks)
       {
         if (chunk.empty())
           continue;
 
+        bool left_joiner = false;
+        bool right_joiner = false;
+        if (_support_prior_joiners) {
+          /* check if prior joiner */
+          if (chunk.length() >= Tokenizer::joiner_marker.length() &&
+              chunk.substr(0, Tokenizer::joiner_marker.length()) == Tokenizer::joiner_marker) {
+            left_joiner = true;
+            chunk = chunk.substr(Tokenizer::joiner_marker.length());
+          }
+
+          if (chunk.length() >= Tokenizer::joiner_marker.length() &&
+              chunk.substr(chunk.length()-Tokenizer::joiner_marker.length()) == Tokenizer::joiner_marker) {
+            right_joiner = true;
+            chunk.erase(chunk.length()-Tokenizer::joiner_marker.length());
+          }
+        }
+
+        if (!_no_substitution)
+          for(size_t i=0; i<special_chars.size(); i++) {
+            const std::string &special_char = special_chars[i];
+            size_t p = 0;
+            while ((p=chunk.find(special_char, p)) != std::string::npos)
+              chunk.replace(p, special_char.length(), substitutes[i]);
+          }
+
         std::vector<std::string> fields = unicode::split_utf8(chunk, ITokenizer::feature_marker);
 
+        size_t p = annotated_tokens.size();
+
         _tokenizeByPlaceholder(fields[0], annotated_tokens, _preserve_placeholders);
 
+        if (left_joiner)
+          annotated_tokens[p].join_left();
```

Comment:
Should this be `p - 1`? If not, a comment should be added because it seems `p` is out of range here.

Reply:
it is p the first token of the new added batch by the `tokenizeByPlaceholder` function. I added a comment, and fixed the right joiner adding that was incorrect.